### PR TITLE
WIP: DBZ-777 PostgreSQL connector rewrite to the standard framework

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-777

@gunnarmorling This is for sneak peek only. Most of the work is done, remaining isssues are toast columns and heartbeats - I am working on it know.
I've tested it with protobuf only.

At the end I've not included the refactorign work you've started with but this will be done as a follow-up as now we better now where potential duplication lies.

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java [hunk 1]
```diff
@@ -396,33 +413,37 @@ private Timer getTableScanLogTimer() {
...
     /**
      * Returns a valid query string for the specified table, either given by the user via snapshot select overrides or
      * defaulting to a statement provided by the DB-specific change event source.
      *
      * @param tableId the table to generate a query for
-     * @return a valid query string
+     * @return a valid query string or empty if table will not be snapshotted
      */
-    private String determineSnapshotSelect(SnapshotContext snapshotContext, TableId tableId) {
+    private Optional<String> determineSnapshotSelect(SnapshotContext snapshotContext, TableId tableId) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java]: Hey @addisonj, this one raised my curiousity -- it's caused by the fact that `Snapshotter` considers snapshot selects as optional. I'm wondering why that is, I'd have said there should always be a snapshot select (filtering of tables is a separate concern of course). So could you shed some light on the reasoning behind this? Thanks!

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                                        connectorConfig,
+                                        schema,
+                                        connection,
+                                        message
+                                )
+                        );
+                })) {
+                    if (offsetContext.hasCompletelyProcessedPosition()) {
+                        dispatcher.dispatchHeartbeatEvent(offsetContext);
+                        pauseNoMessage.pause();
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java]: @jpechane, should the `pause()` call be done in any case, also if the current LSN isn't fully processed yet? I'm wondering whether this creates some sort of "busy spinning" loop otherwise. WDYT?

Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java]: @gunnarmorling Right

File: debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java [hunk 3]
```diff
@@ -57,4 +60,11 @@
      * Signals that a snapshot has been completed, which should reflect in an updated offset state.
      */
     void postSnapshotCompletion();
+
+    /**
+     * Records the name of the table and the timestamp of the last event
+     * @param tableId
+     * @param timestamp
+     */
+    void event(TableId tableId, Instant timestamp);
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java]: I only see this being called once, in the `RelationalSnapshotChangeEventSource`; why not during streaming?

File: debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java [hunk 4]
```diff
@@ -127,4 +144,8 @@ private void emitDeleteRecord(Receiver receiver, TableSchema tableSchema) throws
      * Returns the new row state in case of a CREATE or READ.
      */
     protected abstract Object[] getNewColumnValues();
+
+    protected boolean skipEmptyMessages() {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java]: Why is this needed?

File: debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java (hunk 3 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java]: Beacuse this is ccommon onfly for snapshot. In streaming every connector has its own streaming context metadata.

**Comment by gunnarmorling** [comment]
@jpechane Pushed one more commit and added a few questions. LGTM overall, thanks a lot!

File: debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java (hunk 4 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java]: https://github.com/debezium/debezium/blob/master/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java#L348

File: debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java]: Ok, can we find a more descriptive name then? And can you check with the SQL Server connector, I can't find where source time and table id are advanced during streaming for that one now.

File: debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java (hunk 4 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java]: Ok; a method comment would be good then (general advice for all API/SPI members).

File: debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java (hunk 3 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java]: @gunnarmorling ok, I stand corrected the `evetn` method is used in`io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource.execute(ChangeEventSourceContext)`

Just in PostgresConnectorOffset there is a single method call updating all metadata at once. In case of SqlSetrver there is call of two methods - `event` and `setChangePosition`. So I propose to keep it as is and as a follow-up task to introduce new or extend the `event` method with a parameterized type parameter that would accept `Coordinates` class that will be specific per-connector and contain table, timestampt and other fields.

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java]: Done: https://issues.jboss.org/browse/DBZ-1435

File: debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java (hunk 4 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java]: @gunnarmorling Added

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java (hunk 2 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java]: @gunnarmorling Fixed

File: debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java (hunk 1 above)
Comment (addisonj) [reply debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java]: I think the idea was there might be some tables that you want to stream but you wouldn't want to be including in the initial snapshot for some reason, we have some use cases where we have higher volume tables where old data doesn't matter, so we would prefer not to snapshot it. Obviously, we could just put a query in that would `limit 1` or something so I don't think it would be a huge deal to make it be required.

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java]: Ok, thanks for clarifying, @addisonj. We'll leave it as optional for the time being, may re-consider when merging the current PG-specific `Snapshotter` SPI with the more generic `SnapshotChangeEventSource` contract.

**Comment by gunnarmorling** [comment]
Rebased and applied. Huge thank you to you, @jpechane! 🎉
